### ISSUE TITLE
[3][Keegan-lee] Portfolio Columns improvements

### ISSUE THREAD

**Issue by abitmore** [bitshares/bitshares-ui#2655]
**Is your feature request related to a problem? Please describe.**
Inspired by https://github.com/bitshares/bitshares-ui/issues/2140#issuecomment-489259913, we can have "in orders", "in collateral", "in vesting balances" columns in portfolio table.

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

**Comment by startailcoon** [comment]
Sounds like good ideas, nice.

At the same time I'd like to suggest that we combine the static features to a drop-down menu. The following aren't required to be directly visible to have a function.

- Send, Buy, Deposit, Withdraw, Trade, Borrow, Settle, Burn

This would free up more space and make the dashboard more mobile friendly.

**Comment by Keegan-lee** [comment]
I'd like to claim this.

**Comment by startailcoon** [comment]
@Keegan-lee I'll assign you to this task. Let us know if there's anything you require.

**Comment by Keegan-lee** [comment]
> Sounds like good ideas, nice.
> 
> At the same time I'd like to suggest that we combine the static features to a drop-down menu. The following aren't required to be directly visible to have a function.
> 
> * Send, Buy, Deposit, Withdraw, Trade, Borrow, Settle, Burn
> 
> This would free up more space and make the dashboard more mobile friendly.

Proposing a solution to the above comment.

Combining those "static" functions into a dropdown. 

I propose the header in the table be "Action", where the options are whatever options are valid for that particular asset. In my screenshot. The options for BTS I have available are Buy/Sell/Burn. Those would be the three options in the "Action" dropdown.

See Screenshot for example reference. 

<img width="1280" alt="Screen Shot 2019-05-08 at 12 48 31 PM" src="https://user-images.githubusercontent.com/13334179/57388933-9d9bb480-718f-11e9-865f-4a2f164d62e7.png">

**Comment by startailcoon** [comment]
Action sounds like a good name for the title. These actions should use the style-guide drop down menu and only be populated with the options available, indeed. I would opt that `Hide` should stay at the far right as today.

BTS for example here, doesn't have withdraw, trade, borrow, settle actions.

Make sure you catch all places in code that this needs to be changed/added. We have the visible list, as well as the hidden list.

**Comment by abitmore** [comment]
Can we make the "put actions to a drop down" a responsive design? For wide screen users, it's not wise to hide them by default, because that means additional clicks.

**Comment by Keegan-lee** [comment]
> Action sounds like a good name for the title. These actions should use the style-guide drop down menu and only be populated with the options available, indeed. I would opt that `Hide` should stay at the far right as today.
> 
> BTS for example here, doesn't have withdraw, trade, borrow, settle actions.
> 
> Make sure you catch all places in code that this needs to be changed/added. We have the visible list, as well as the hidden list.

Thanks for the clarification. Working on this now. I'll let you know if there are any additional difficulties.

**Comment by sschiessl-bcp** [comment]
> Can we make the "put actions to a drop down" a responsive design? For wide screen users, it's not wise to hide them by default, because that means additional clicks.

I'm with abit here, responsiveness is key. 

On a side note: With the redesign going on I'd rather see the core of the wallet improved and not the UI

**Comment by Keegan-lee** [comment]
I don't think the actionDropdown is good UX. See the screenshot below. Once a person has selected the action they want to take, then what? Do they need to then click a confirm button of sorts to actually take the action? This introduces another click into the flow. 

Thoughts or suggestions welcome.

<img width="1280" alt="Screen Shot 2019-05-14 at 11 19 31 PM" src="https://user-images.githubusercontent.com/13334179/57744224-c3005500-769e-11e9-90f9-3a22f81898b9.png">

**Comment by sschiessl-bcp** [comment]
Initial issue was about "in orders", "in collateral", "in vesting balances". Do you have a suggestion?

Maybe @ahdigital has an idea?

Basically following problem: We want to summarize the actions (withdraw, deposit, buy, maybe trade and send) and also add more information to where balances are (orders, collateral, vesting and normal balances). And that all should fit in one row. Since this is a central piece it needs to be pretty!

I'd suggest a bounty of 4 for UX.

**Comment by Keegan-lee** [comment]
Right there are two parts to this issue. 

1. The addition of "in orders", "in collateral", and "in vesting balances". 

2. The removal / tidying / combining into a dropdown menu. I still don't think this is a reasonable solution as the UX would suffer IMO. Open to suggestions here.

**Comment by ahdigital** [comment]
I agree that there are two issues at play here.

Regarding the second point - Send, Buy, Deposit, Withdraw, Trade, Borrow, Settle and Burn are all triggers to begin a new user journey, because each opens a modal window with a different form to complete. It's not just a 'confirmation' step.

The usage of a drop down comes down to how often each action is used. Are users acting on each row in quick succession, meaning an extra click reduces efficiency, or are these actions less common, and actually the space saving benefit is of more value? I don't see the harm in using the drop downs, myself.

We might want to consider though, if the actions column work is part of a bigger story, that is: how to deal with tables that have many columns. Would some column selector functionality also deliver value, for example?

**Comment by froooze** [comment]
The TRADE button is the only important one. Send, BUY ... can be put in drop down menu.

**Comment by startailcoon** [comment]
The actions we have that we can do something with are the following

- Send. Should be visible all the time, as send is an important part.
- Hide. Should be visible all the time, as you should easily be able to hide.
- Trade. Should be visible all the time, as trade is an important part.
- Buy. Most likely not used widely as it's a TradeBlock integration buy.
- Deposit. Somewhat used probably, but would propose to be in the menu.
- Withdraw. Probably used more than deposit, but would propose to be in the menu.
- Borrow. Somewhat less used, advanced feature. Also visible on the market.
- Settle. Somewhat less used, advanced feature. Not visible on the market.
- Burn. Likely not used much, advanced feature for asset creators.

So, Send, Trade and Hide should be visible. The rest could probably go in a menu. When a user clicks on the menu entry all these actions opens a modal, so it would still be fluid for the user.

This is just to make place for the more text/space demanding "In Orders", "In Collateral" and "In Vesting Balance" data.

**Comment by sschiessl-bcp** [comment]
> Right there are two parts to this issue.
> 
> 1. The addition of "in orders", "in collateral", and "in vesting balances".
> 2. The removal / tidying / combining into a dropdown menu. I still don't think this is a reasonable solution as the UX would suffer IMO. Open to suggestions here.
> 
> @startailcoon

Please explore options to allow editing of visible columns through antd table and report back in here.

**Comment by Keegan-lee** [comment]
https://ant.design/components/table/#components-table-demo-fixed-columns

I think the fixed-columns would be a good way to accomplish what we're looking for. A subsection of the table would be scrollable, allowing for the fitting of every action we want to be there, all the while being mindful of the UI/UX and not having too many items in the table. 

Thoughts?

@sschiessl-bcp

**Comment by sschiessl-bcp** [comment]
I was thinking a more sophisticated approach where we introduce a higher-level configurable table component that allows configuration of columns (e.g. as done here https://stackoverflow.com/questions/53952293/antd-ui-table-dynamically-add-delete-columns)

Example (not meant as a mockup):
![image](https://user-images.githubusercontent.com/33128181/59012072-d1bdd000-8836-11e9-85a9-341d2a3b1642.png)
I imagine a customize icon on the top right of the table that allows to select which columns to show

**Comment by Keegan-lee** [comment]
@sschiessl-bcp - I do think this would be better suited in a different issue. The original issue is completed, and I'm 4-5 hours into the issue. I like your this suggestion though, thank you for the articulation. I'd be happy to take on the new issue once its created.

**Comment by startailcoon** [comment]
I would suggest that we merge the work today with the new fields completed by @Keegan-lee to finalize the current issue scope. A new component, as you suggest @sschiessl-bcp, should be made into a new issue and worked on there instead. It would allow the current task to finish and the new one to form before we start its work.

@sschiessl-bcp please create a new issue for the advanced table ideas.

**Comment by sschiessl-bcp** [comment]
I'd rather not change the UX twice towards the user. If @Keegan-lee is fine with it I would rather hold this issue until #2761 is completed and use that new component here right away.

**Comment by abitmore** [comment]
@sschiessl-bcp I still think we can add the new columns first. Buttons are another issue, so we're not changing same thing twice.
